src/dex.ts

```diff
@@ -0,0 +1,456 @@
+    invariant(amountIn > 0n, "amount in must be positive");
+    invariant(minimumLPReceived > 0n, "minimum LP received must be positive");
+    const lucid = await this.getLucidInstance();
+    lucid.selectWalletFrom({ address: sender });
+    const orderAssets: Assets = { [assetIn]: amountIn };
+    const utxos = await lucid.utxosAt(sender);
+    const { batcherFee, reductionAssets } = this.calculateBatcherFee(
+      utxos,
+      orderAssets
+    );
+    if (orderAssets["lovelace"]) {
+      orderAssets["lovelace"] += FIXED_DEPOSIT_ADA + batcherFee;
+    }
+    const datum: OrderDatum = {
+      sender: sender,
+      receiver: sender,
+      receiverDatumHash: undefined,
+      step: {
+        type: OrderStepType.ZAP_IN,
+        desiredAsset: assetOut,
+        minimumLP: minimumLPReceived,
+      },
+      batcherFee: batcherFee,
+      depositADA: FIXED_DEPOSIT_ADA,
+    };
+
+    return await lucid
+      .newTx()
+      .payToContract(
+        ORDER_BASE_ADDRESS[this.networkId],
+        OrderDatum.toCborHex(datum),
+        orderAssets
+      )
+      .payToAddress(sender, reductionAssets)
+      .addSigner(sender)
+      .attachMetadata(674, { msg: [MetadataMessage.ZAP_IN_ORDER] })
+      .complete();
+  }
+
+  async buildDepositTx(options: BuildDepositTxOptions): Promise<TxComplete> {
+    const { sender, assetA, assetB, amountA, amountB, minimumLPReceived } =
+      options;
+    invariant(amountA > 0n && amountB > 0n, "amount must be positive");
+    invariant(minimumLPReceived > 0n, "minimum LP received must be positive");
+    const lucid = await this.getLucidInstance();
+    lucid.selectWalletFrom({ address: sender });
+    const orderAssets = {
+      [assetA]: amountA,
+      [assetB]: amountB,
+    };
+    const utxos = await lucid.utxosAt(sender);
+    const { batcherFee, reductionAssets } = this.calculateBatcherFee(
+      utxos,
+      orderAssets
+    );
+    if (orderAssets["lovelace"]) {
+      orderAssets["lovelace"] += FIXED_DEPOSIT_ADA + batcherFee;
+    }
+    const datum: OrderDatum = {
+      sender: sender,
+      receiver: sender,
+      receiverDatumHash: undefined,
+      step: {
+        type: OrderStepType.DEPOSIT,
+        minimumLP: minimumLPReceived,
+      },
+      batcherFee: batcherFee,
+      depositADA: FIXED_DEPOSIT_ADA,
+    };
+    return await lucid
+      .newTx()
+      .payToContract(
+        ORDER_BASE_ADDRESS[this.networkId],
+        OrderDatum.toCborHex(datum),
+        orderAssets
+      )
+      .payToAddress(sender, reductionAssets)
+      .addSigner(sender)
+      .attachMetadata(674, { msg: [MetadataMessage.DEPOSIT_ORDER] })
+      .complete();
+  }
+
+  async buildCancelOrder(
+    options: BuildCancelOrderOptions
+  ): Promise<TxComplete> {
+    const { orderTxId, sender } = options;
+    const orderUtxo = blockfrostToLucidUtxo(
+      await this.blockfrostAdapter.getOrderUtxoByTxId(orderTxId)
+    );
+    const lucid = await this.getLucidInstance();
```

Comment:
just pass `Lucid.Utxo` to this function
we let user chooses the best way to query and store order utxos